heartbeat/componentHandler/heartbeatHandler.go

```diff
@@ -0,0 +1,274 @@
+		arg.KeyGenerator,
+		arg.Marshalizer,
+		arg.PeerShardMapper,
+	)
+	if err != nil {
+		return err
+	}
+
+	heartbeatStorer, err := heartbeatStorage.NewHeartbeatDbStorer(arg.Storer, arg.Marshalizer)
+	if err != nil {
+		return err
+	}
+
+	timer := &process.RealTimer{}
+	netInputMarshalizer := arg.Marshalizer
+	if arg.SizeCheckDelta > 0 {
+		netInputMarshalizer = marshal.NewSizeCheckUnmarshalizer(arg.Marshalizer, arg.SizeCheckDelta)
+	}
+
+	allValidators, _, _ := hbh.getLatestValidators()
+	pubKeysMap := make(map[uint32][]string)
+	for shardID, valsInShard := range allValidators {
+		for _, val := range valsInShard {
+			pubKeysMap[shardID] = append(pubKeysMap[shardID], string(val.PublicKey))
+		}
+	}
+
+	argMonitor := process.ArgHeartbeatMonitor{
+		Marshalizer:                        netInputMarshalizer,
+		MaxDurationPeerUnresponsive:        time.Second * time.Duration(arg.HeartbeatConfig.DurationInSecToConsiderUnresponsive),
+		PubKeysMap:                         pubKeysMap,
+		GenesisTime:                        arg.GenesisTime,
+		MessageHandler:                     heartBeatMsgProcessor,
+		Storer:                             heartbeatStorer,
+		PeerTypeProvider:                   peerTypeProvider,
+		Timer:                              timer,
+		AntifloodHandler:                   arg.AntifloodHandler,
+		HardforkTrigger:                    arg.HardforkTrigger,
+		PeerBlackListHandler:               arg.PeerBlackListHandler,
+		ValidatorPubkeyConverter:           arg.ValidatorPubkeyConverter,
+		HbmiRefreshIntervalInSec:           arg.HeartbeatConfig.HbmiRefreshIntervalInSec,
+		HideInactiveValidatorIntervalInSec: arg.HeartbeatConfig.HideInactiveValidatorIntervalInSec,
+	}
+	hbh.monitor, err = process.NewMonitor(argMonitor)
+	if err != nil {
+		return err
+	}
+
+	log.Debug("heartbeat's monitor component has been instantiated")
+
+	err = hbh.monitor.SetAppStatusHandler(arg.AppStatusHandler)
+	if err != nil {
+		return err
+	}
+
+	err = arg.Messenger.RegisterMessageProcessor(core.HeartbeatTopic, hbh.monitor)
+	if err != nil {
+		return err
+	}
+
+	var ctx context.Context
+	ctx, hbh.cancelFunc = context.WithCancel(context.Background())
+	go hbh.startSendingHeartbeats(ctx)
+
+	return nil
+}
+
+func (hbh *HeartbeatHandler) getLatestValidators() (map[uint32][]*state.ValidatorInfo, map[string]*state.ValidatorApiResponse, error) {
+	latestHash, err := hbh.arg.ValidatorStatistics.RootHash()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	validators, err := hbh.arg.ValidatorStatistics.GetValidatorInfoForRootHash(latestHash)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return validators, nil, nil
+}
+
+func (hbh *HeartbeatHandler) startSendingHeartbeats(ctx context.Context) {
+	r := rand.New(rand.NewSource(time.Now().Unix()))
+	cfg := hbh.arg.HeartbeatConfig
+
+	log.Debug("heartbeat's endless sending go routine started")
+
+	for {
+		diffSeconds := cfg.MaxTimeToWaitBetweenBroadcastsInSec - cfg.MinTimeToWaitBetweenBroadcastsInSec
```

Comment:
no need to compute timeToWait in every iteration, as it should always be the same, can be done outside of the loop

Reply:
agree, changed